### ISSUE TITLE
Can't add files with names which contain "/" and "." and "~" in object name

### ISSUE THREAD

**Issue by Gr1g0ry** [GoogleCloudPlatform/gcloud-node#1340]
When I use object names with __dot__, __slash__ and __tilde__ together I receive error. 

Next is mocha test: 

```javascript
"use strict";
var config = require('config');
var gcloud = require('gcloud');
var gcs = gcloud.storage({
  keyFilename: config.gcloud.keyFilename,
  projectId:   config.gcloud.projectId
});
var myBucket = gcs.bucket(config.buckets.serviceTags.name);

describe('Zero test', function () {
  it('should save file without errors', function (done) {
    myBucket.file('type~user/user.00000000-0000-0000-0000-000000000000')
      .save('some text', {metadata: {contentType: 'text/plain' }}, function (err) { done(err); });
  });
});
```

returns error:
>      Error: Not Found
>       at _stream_readable.js:908:16

**Comment by stephenplusplus** [comment]
I couldn't reproduce with the following script:

```js
'use strict'

var gcloud = require('gcloud')
var gcs = gcloud.storage()
var bucket = gcs.bucket(process.env.BUCKET)
var file = bucket.file('type~user/user.00000000-0000-0000-0000-000000000000')

file.save('test', console.log)
```

Is there any chance the error is coming from a separate call? Maybe in a `before` hook?

**Comment by Gr1g0ry** [comment]
I'm sure that problem in test. I have the same problem also in node

```javascript
"use strict";
var config = require('config');
var gcloud = require('gcloud');
var gcs = gcloud.storage({
  keyFilename: config.gcloud.keyFilename,
  projectId:   config.gcloud.projectId
});
var myBucket = gcs.bucket(config.buckets.serviceTags.name);
var file = myBucket.file('type~user/user.00000000-0000-0000-0000-000000000000');

file.save('some text', {metadata: {contentType: 'text/plain' }}, function (err) { if (err) {throw err; } });
```

**Comment by stephenplusplus** [comment]
I can reproduce when the bucket doesn't exist. Could you check that the bucket exists?

```js
bucket.exists(console.log)
```

**Comment by Gr1g0ry** [comment]
Yes bucket exists. For reproducing you need some of your _bucket_ and correspondent _key.json_ file.
I'm sure that problem in name because next code is successful:

```javascript
"use strict";
var config = require('config');
var gcloud = require('gcloud');
var gcs = gcloud.storage({
  keyFilename: config.gcloud.keyFilename,
  projectId:   config.gcloud.projectId
});
var myBucket = gcs.bucket(config.buckets.serviceTags.name);
var file = myBucket.file('nice/name');

file.save('some text', {metadata: {contentType: 'text/plain' }}, function (err) { if (err) {throw err; } });
```

**Comment by stephenplusplus** [comment]
What OS are you running?

**Comment by Gr1g0ry** [comment]
Ubuntu 14.04

**Comment by stephenplusplus** [comment]
Can you try:

```js
file.save('test', {
  resumable: false,
  metadata: {
    contentType: 'text/plain'
  }
}, console.log)
```

**Comment by Gr1g0ry** [comment]
Ok. The problem was solved.

**Comment by Gr1g0ry** [comment]
Should I close issue?

**Comment by stephenplusplus** [comment]
Glad it was resolved. Let's keep it open for a little longer. I want to dig into this later to figure out exactly what might have happened.

**Comment by Gr1g0ry** [comment]
Ok.

**Comment by stephenplusplus** [comment]
I think I've found a resolution. That was an error from `gcs-resumable-upload`, the module we use to handle resumable file uploads. In the beginning part of trying to resume an upload, it asks the server how much data it already has, so it knows at what point in the file it should start sending data from. The response to that was a 404, because of a caching issue. To track state, that module stores a piece of data in a config file. The file you were uploading must have been uploaded previously, and either it failed or it was improperly cleaned up after the file was uploaded.

The fix I just released will automatically recognize that scenario (caching error), erase the old entry from the cache, and start a new resumable upload.

Thanks for catching this, it will hopefully save a lot of confusion for users going forward!
